Changes to Covid19-data Repo by Hongtao

hongtaoh opened the pull request:
I manipulated the original data directly without going through the [covid-19 data repo](https://github.com/covid19-data/covid19-data). [Here](https://observablehq.com/@hongtaoh/day-46-2020-10-08) is the notebook detailing how I manipulated the data on ObservableHQ. The [most updated dataset](https://ourworldindata.org/coronavirus-source-data) from Our World in Data (OWID) already contains country codes and population size, so I am directly using them. 

However, OWID dataset does not have the information on country's region classification as you (and the data from World Bank) did. It only specifies in which continent each country resides. I think world regions makes more sense than simply continents. Therefore, I assigned a region to each country using Wrold Bank's data. There are ten countries / areas whose data were not found in the World Bank data. For those, I used data from [ISO-3166-Countries-with-Regional-Codes](https://github.com/hongtaoh/covid19-data/blob/master/data_sources/metadata/worldbank/country_metadata.csv), as the [README](https://github.com/hongtaoh/covid19-data#readme) file has detailed.

The only problem is that in your original visualizations, "ERI" (Eritrea) was labelled as "East & North Africa" in the [original data](https://observablehq.com/@yy/covid-19-fatality-rate#data), but it was classified as in "Sub-Sahara Africa" in World Bank data. Also, "ERI" is the only country with the label of "East & North Africa". So I changed its label to "Sub-Sahara Africa" and deleted the label of "East & North Africa" in all visualizations. 

"BES" was the only one labeled as "America & Caribbean" in the original data. I changed it to "Latin America & Caribbean" according to [ISO-3166-Countries-with-Regional-Codes](https://github.com/hongtaoh/covid19-data/blob/master/data_sources/metadata/worldbank/country_metadata.csv). I then deleted the label of "America & Caribbean" in all visualizations.

I changed the starting date for [COVID-19 Confirmed vs. New cases](https://observablehq.com/d/87e6003e4f4aab00) to be Dec. 31th, 2019, because it presented a whole picture of the pandemic from the earliest start. To do this, I modified the `scripts/nyt_state_data.py` such that data for all US states also started from 2019-12-31; otherwise, there will be errors in the visualization. 

I move all the files not in use currently to the folder of deprecated. 

I didn't know to use Snakemake. Since right now, there are only two Python scripts, I managed to automate the data updating process using this [bash script](https://github.com/hongtaoh/covid19-data/blob/master/deploy.sh).

yy commented on .gitignore:
```diff
@@ -1,137 +0,0 @@
-*.csv
```
Please keep the gitignore file

hongtaoh replied:
.gitignore file was just added.

yy commented on .gitmodules:
```diff
@@ -4,3 +4,9 @@
+[submodule "data_sources/metadata/ISO-3166-Countries-with-Regional-Codes"]
+	path = data_sources/metadata/ISO-3166-Countries-with-Regional-Codes
+	url = https://github.com/lukes/ISO-3166-Countries-with-Regional-Codes
+[submodule "ISO-3166-Countries-with-Regional-Codes"]
```
same repo has been duplicated.

hongtaoh commented:
Professor YY, I've made relevant changes according to your suggestions, added the `.gitignore` and edited the `gitmodules`. If I did it wrong, please let me know. It's my first time editing `gitmodules`. 

Also, I've added the script of [`prepare_owid_viz_data.py`](`https://github.com/hongtaoh/covid19-data/blob/master/scripts/), so that we don't need to go through the Observable notebook for data manipulation. Something worth noticing in the script: when we normalize each country/area with the same date range, it's better not to simply `fillna(0)`. This is because there are some NaN within the data, even before normalization. For example, for data of AFG, there were no data for March 3rd to March 7, and for data of ABW, there were no data for March 13 to March 19. What I did is for each country, if the data for December 31st, 2019 is NaN, I assigned a 0. Then, I filled the NaN via forward filling. 

There are still some minor issues with the script: 

1. The script will give me a warning: `A value is trying to be set on a copy of a slice from a DataFrame`. I'll try to fix it later;
2. The forward filling for NA bring an issue: as I mentioned in our earlier chat, four places do not have the most updated data, so there are NA at the end of the data of these places. However, Forward filling will fill these NAs as well, which is not what we expect. I'll try fixing it later. 

Something I need to do: I'll have a look at whether the script of `deploy.sh` can run by itself everyday, so that I don't need to submit PR everyday later.

hongtaoh reviewed (commented):
Professor YY, I've made relevant changes according to your suggestions, added the .gitignore and edited the gitmodules. If I did it wrong, please let me know. It's my first time editing gitmodules.

Also, I've added the script of [`prepare_owid_viz_data.py`](https://github.com/hongtaoh/covid19-data/blob/master/scripts/prepare_owid_viz_data.py), so that we don't need to go through the Observable notebook for data manipulation. Something worth noticing in the script: when we normalize each country/area with the same date range, it's better not to simply fillna(0). This is because there are some `NaN` within the data, even before normalization. For example, for data of AFG, there were no data for March 3rd to March 7, and for data of ABW, there were no data for March 13 to March 19. What I did is for each country, if the data for December 31st, 2019 is `NaN`, I assigned a 0. Then, I filled the `NaN` via forward filling.

There are still some minor issues with the script:

1. The script will give me a warning: A value is trying to be set on a copy of a slice from a DataFrame. I'll try to fix it later;
2. The forward filling for NA bring an issue: as I mentioned in our earlier chat, four places do not have the most updated data, so there are NA at the end of the data of these places. However, Forward filling will fill these `NaN`s as well, which is not expected. I'll try fixing it later.

Something I need to do: I'll have a look at whether the script of deploy.sh can run by itself everyday, so that I don't need to submit PR everyday later.

yy commented on README.md:
```diff
@@ -8,26 +8,31 @@ This repository hosts workflows to process several data sources and cleaned data
 ### Historical case data for visualizations
 
-- [`cntry_stat_owid.json`](https://github.com/yy/covid19-data/blob/master/output/cntry_stat_owid.json): ECDC historical data merged with Worldbank's country metadata and ISO 3166-1 Alpha-3 data. Used in:
+- [`cntry_stat_owid.json`](https://github.com/hongtaoh/covid19-data/blob/master/output/cntry_stat_owid.json): ECDC historical data merged with Worldbank's classification of world regions. Used in:
```
if we merge this, I think we want to have `covid19-data/covid19-data`

hongtaoh replied:
That's right. I have just edited the `README.md`.

yy commented on deprecated/notebooks/.ipynb_checkpoints/HT_OWID_manipulation-checkpoint.ipynb:
```diff
@@ -0,0 +1,273 @@
+{
```
add `.ipynb_checkpoints` to gitignore and remove these temporary files.

hongtaoh replied:
`.ipynb_checkpoints` was just added to `.gitignore`. Do you mean these two `untitled.ipynb` within [`deprecated/notebooks`](https://github.com/hongtaoh/covid19-data/tree/master/deprecated/notebooks)? I am not quite sure which files to remove.

yy replied:
files within the `.ipynb_checkpoints` folder, not within notebooks folder.

hongtaoh replied:
Got it! I just have removed the `ipynb_checkpoints` folder within `deprecated/notebooks`.  Sorry for all mistakes made.

hongtaoh commented:
Dear Professor YY:

I've just updated the [data manipulation script](https://github.com/hongtaoh/covid19-data/blob/master/scripts/prepare_owid_viz_data.py). I fixed some bugs in the previous script. The four places ('SWE', 'ESP', 'BEL', 'HKG') that fall behind now have 'null' in dates that are not updated yet. 

The problem about `last_index` I mentioned in the Spreading Trend notebook was solved automatically. This is because the above four places do have "index" for dates note updated yet; it's just that the data is 'null'. So I don't need to make `last_index` a variable. 

I saw that the Covid19-data repo is still not updated with my proposed changes. Is there anything to be done on my part? I had very few experiences with submitting PR before, so I am not sure what I need to do next. 

Another thing is that although the covid19-dashboard is updated, the webpage (https://yyahn.com/covid19/) on your website did not reflect the changes. I don't know how to update it. Maybe you need to reset the "setting" in the repo. 

My final question: is it possible for a repository to run the `deploy.sh` script by itself. If not, I need to run the script everyday and submit PR. Which is fine with me, though.
